src/cuda/Framework/EDProducer.h

```diff
@@ -27,27 +34,157 @@ namespace edm {
+      if (events.size() > statesSize_) {
+        statesSize_ = events.size();
+        states_ = std::make_unique<AsyncState[]>(statesSize_);
+      }
+      for (size_t i = 0; i < events.size(); ++i) {
+        acquire(events[i], eventSetup, holder, states_[i]);
+      }
     }
 
-    void doProduce(Event& event, EventSetup const& eventSetup) { produce(event, eventSetup); }
+    virtual void acquire(Event const& event,
+                         EventSetup const& eventSetup,
+                         WaitingTaskWithArenaHolder holder,
+                         AsyncState& state) const = 0;
+
+    void doProduce(EventRange events, EventSetup const& eventSetup) {
+      for (size_t i = 0; i < events.size(); ++i) {
+        produce(events[i], eventSetup, states_[i]);
+      }
+    }
+
+    virtual void produce(Event& event, EventSetup const& eventSetup, AsyncState& state) = 0;
+
+    void doEndJob() { endJob(); }
+
+    virtual void endJob() {}
+
+  private:
+    size_t statesSize_ = 0;
+    std::unique_ptr<AsyncState[]> states_;
+  };
+
+  template <>
+  class EDProducerExternalWork<void> {
+  public:
+    EDProducerExternalWork() = default;
+    virtual ~EDProducerExternalWork() = default;
+
+    bool hasAcquire() const { return true; }
+
+    void doAcquire(ConstEventRange events, EventSetup const& eventSetup, WaitingTaskWithArenaHolder holder) {
+      for (Event const& event : events) {
+        acquire(event, eventSetup, holder);
+      }
+    }
+
+    virtual void acquire(Event const& event, EventSetup const& eventSetup, WaitingTaskWithArenaHolder holder) const = 0;
+
+    void doProduce(EventRange events, EventSetup const& eventSetup) {
+      for (Event& event : events) {
+        produce(event, eventSetup);
+      }
+    }
 
-    virtual void acquire(Event const& event, EventSetup const& eventSetup, WaitingTaskWithArenaHolder holder) = 0;
     virtual void produce(Event& event, EventSetup const& eventSetup) = 0;
 
     void doEndJob() { endJob(); }
+
+    virtual void endJob() {}
+
+  private:
+  };
+
+  template <typename T = void>
+  class EDBatchingProducerExternalWork {
+  public:
+    using AsyncState = T;
+
+    EDBatchingProducerExternalWork() = default;
+    virtual ~EDBatchingProducerExternalWork() = default;
+
+    bool hasAcquire() const { return true; }
+
+    void doAcquire(ConstEventRange events, EventSetup const& eventSetup, WaitingTaskWithArenaHolder holder) {
+      acquire(events, eventSetup, holder, state_);
+    }
+
+    virtual void acquire(ConstEventRange events,
+                         EventSetup const& eventSetup,
+                         WaitingTaskWithArenaHolder holder,
+                         AsyncState& state) const = 0;
+
+    void doProduce(EventRange events, EventSetup const& eventSetup) { produce(events, eventSetup, state_); }
+
+    virtual void produce(EventRange events, EventSetup const& eventSetup, AsyncState& states) = 0;
```

Comment:
Really minor, but is this a typo?
```suggestion
    virtual void produce(EventRange events, EventSetup const& eventSetup, AsyncState& state) = 0;
```

Reply:
I've used both spellings while prototyping, so I don't mind either way: this should capture the "state" or "states" of multiple events:
  - `state` is more consistent with `AsyncState`
  - `states` is more consistent with `events`

Re-thinking a bit more... maybe `state` is more correct, since it is one object, not *e.g.* a `vector<AsyncState>`.